### ISSUE TITLE
Sign In cancelled when app go to background

### ISSUE THREAD

**Issue by Danaru87** [aboutyou/dart_packages#162]
**Flutter** : 1.22.0
**Android**: 10
**sign_in_with_apple**: ^2.5.4

We create an application using the Apple Sign In plugin.

When the web view asks for an Apple ID and password, two-factor authentication sends a code via SMS.

But, going to the SMS application to read the code, going back to the Flutter application causes an exception of type: SignInWithAppleAuthorizationError (AuthorizationErrorCode.canceled, The user closed the Custom Tab).

**Comment by tp** [comment]
@Danaru87 Can you provide a screen recording of how the switch towards the SMS app happens?

I would only expect this on resource constrained devices when the browser activity gets closed.

**Comment by Danaru87** [comment]
Flutter: 1.22.3
sign_in_with_apple: ^2.5.4
Device: Motorola One Vision
Android: 10

video:


![bug apple sign in android](https://user-images.githubusercontent.com/5873203/98663420-604a5100-2349-11eb-9d13-b4e27572e316.gif)


error message in console:
```
E/flutter (22882): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: SignInWithAppleAuthorizationError(AuthorizationErrorCode.canceled, The user closed the Custom Tab)
E/flutter (22882): #0      SignInWithApple._signInWithAppleAndroid (package:sign_in_with_apple/src/sign_in_with_apple.dart:239:7)
E/flutter (22882): <asynchronous suspension>
E/flutter (22882): #1      SignInWithApple.getAppleIDCredential (package:sign_in_with_apple/src/sign_in_with_apple.dart:107:14)
E/flutter (22882): #2      _ProviderButtonsWidget.build.<anonymous closure> (package:spheery/provider_choice/provider_choice_screen.dart:95:58)
E/flutter (22882): #3      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:993:19)
E/flutter (22882): #4      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:1111:38)
E/flutter (22882): #5      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:183:24)
E/flutter (22882): #6      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:598:11)
E/flutter (22882): #7      BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:287:5)
E/flutter (22882): #8      BaseTapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:259:7)
E/flutter (22882): #9      GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:157:27)
E/flutter (22882): #10     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:362:20)
E/flutter (22882): #11     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:338:22)
E/flutter (22882): #12     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:267:11)
E/flutter (22882): #13     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:295:7)
E/flutter (22882): #14     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:240:7)
E/flutter (22882): #15     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:213:7)
E/flutter (22882): #16     _rootRunUnary (dart:async/zone.dart:1206:13)
E/flutter (22882): #17     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter (22882): #18     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
E/flutter (22882): #19     _invoke1 (dart:ui/hooks.dart:265:10)
E/flutter (22882): #20     _dispatchPointerDataPacket (dart:ui/hooks.dart:174:5)
E/flutter (22882): 
I/OpenGLRenderer(22882): Davey! duration=35435ms; Flags=1, IntendedVsync=8683343831608, Vsync=8683343831608, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=8683351226156, AnimationStart=8683351273272, PerformTraversalsStart=8683351276887, DrawStart=8718761061095, SyncQueued=8718763847057, SyncStart=8718764651634, IssueDrawCommandsStart=8718764858172, SwapBuffers=8718774698827, FrameCompleted=8718780389558, DequeueBufferDuration=2965000, QueueBufferDuration=1945000, 
```

**Comment by Mikelcastrolads** [comment]
Hi @nikhilmufc7 

Are you able to fix the issue? I'm having a problem signing in using android. It redirects to my app on playstore and when I click open it says:

> Unhandled Exception: SignInWithAppleAuthorizationError(AuthorizationErrorCode.canceled, The user closed the Custom Tab)

**Comment by 571cky** [comment]
Igual por aca.

**Comment by HenriBeck** [comment]
> Are you able to fix the issue? I'm having a problem signing in using android. It redirects to my app on playstore and when I click open it says:

@nikhilmufc7 This sounds like your redirect isn't working as expected when you go into the play store.

Please double-check your package identifier and `AndroidManifest.xml` file.

**Comment by liri2006** [comment]
We face the same issue - tested on multiple Android devices, including latest Pixel. After app is resumed from the background browser is closed.

**Comment by istvmunkacsi** [comment]
I also faced a described issue on the emulator and different real devices. 

I think this is happening because "FLAG_ACTIVITY_NO_HISTORY" is being added to CustomTabsIntent (this flag specifically asks to not add a new activity to the history stack). Removing the line of code responsible for that resolved the issue for me. 

I couldn't find any issues during testing after removing the flag, but maybe I just didn't test well enough 😃 @tp Could you please share why this flag was used?

**Comment by tp** [comment]
@istvmunkacsi Hmm, I don't recall the exact reason for that flag, but there were multiple issues with the activities on Android:

- in some configurations the custom tab would be popped / dismissed when the user backgrounded the app and then brought it back to focus by tapping the app icon (this is not ideal for our use case, we would rather have the app stay as it was in that case, just like on iOS).
- in some configuration we weren't able to get the custom tab dismissed with the incoming deep link

I wish I had better notes from the implementation to be more specific here. Overall I think there might be better approaches to this these days (maybe only for never API levels?), so might be worth investigating further if that can help with another issue.

**Comment by istvmunkacsi** [comment]
Thanks for the answer @tp. The first issue you described is exactly the issue caused by adding "FLAG_ACTIVITY_NO_HISTORY" to CustomTabsIntent. CustomTabs activity is closed in every scenario when the app is brought back from the background (by clicking the app's icon as you described, or just selecting it from "currently opened apps" as shown by @Danaru87)

But you are right - removing this flag will result in 2nd issue you described. To make both scenarios work as expected I propose: 

- remove adding "FLAG_ACTIVITY_NO_HISTORY" to CustomTabsIntent on the package side
- add to the package's documentation, that to avoid issues with deep links 3rd party app needs to add `android:launchMode="singleTop"` property in the app's AndroidManifest for the activity it is aiming to return to. This should resolve the 2nd issue without creating 1st. Source: https://bugs.chromium.org/p/chromium/issues/detail?id=545446

**Comment by tp** [comment]
@istvmunkacsi Good point, the launch mode of the app is also relevant here. We should definitely explain that. Though if I recall correctly we use `singleTask` in our main app. But maybe both `singleTask` and `singleTop` lead to the desired result?

**Comment by tp** [comment]
> Could we expect a fix in the near time?

@istvmunkacsi I'll look into it this week. I suppose ("fear") some showstopper might show up though, as I recall that we've been testing a bunch of configurations until we got it working for the main flows in our app (without breaking other integrations etc.). As far as I recall we even had to make a compromise on the "custom tab is dismissed when tapping the app icon again" (but not the app switcher).

Anyway, it will surely be useful to dig into this again and at the very least gain some more clarity (and documentation) on which flags result in what behavior (and maybe other apps will choose different trade-offs).

**Comment by istvmunkacsi** [comment]
@tp Hmm, there shouldn't be a case when "custom tab is dismissed when tapping the app icon again", but "custom tab is **not** dismissed when returning back to it using app switcher". CustomTab is always closed when the app returns from the background regardless of the scenario. But take into consideration, that when you press "Overview" button on "built-in" navigation bar (the one that opens app switcher) - currently active app is not being sent to the background immediately. Only when another app is selected, or "Home" is invoked (by tapping a "Home" button or swipe gesture). So if: 

- the app is opened -> Sign-in CustomTab is shown -> "Overview" button is pressed -> the same app is selected - CustomTab won't be closed (regardless of the activity or Intent flags) because app was not sent to background. But if:
- app is opened -> Sign-in CustomTab is shown -> _"Home" is invoked_ or _"Overview" button is pressed & another app is selected_ ->  app with opened CustomTab is re-opened (in any way - by tapping app icon or with app switcher) - CustomTab will be closed if `FLAG_ACTIVITY_NO_HISTORY` is being added to CustomTabsIntent as it is now.

**Comment by tp** [comment]
@istvmunkacsi I gave this a spin locally and indeed removing the `NO_HISTORY` flag works fine with both `singleTop` and `singleTask`. Though they behave slightly differently. (See README for further details)

I published a pre-release version for you all to test: https://pub.dev/packages/sign_in_with_apple/versions/5.0.0-dev.1

Unfortunately I couldn't verify this with the app in `example/` as in the end I was redirected to the Play Store instead of the app after sign-in. I suppose somewhere in the chain we have a different APP-ID coded in the sample code and I will need to adapt that for the example app. But I'll verify that myself as well before publishing.

Hope this alleviates the issue for you, if you find any other flow that doesn't work as intended, let me know :)

**Comment by naamapps** [comment]
@tp 
What is the recommended launch mode after the change? 
SingleTask or SingleTop?

**Comment by istvmunkacsi** [comment]
Thank you @tp

**Comment by tp** [comment]
@naamapps For the Chrome Custom Tab specifically I think `singleTop` is best, as that ensure that the browser is kept on top no matter how one switches out and back into the app.

<img width="1132" alt="CleanShot 2022-12-12 at 10 00 44@2x" src="https://user-images.githubusercontent.com/52437/207004281-92ebe348-c88b-47cd-aa5e-249eac8836ca.png">


But I am not 100% clear if there would be cases where that has undesirable side-effects. The first thing which comes to mind is deep-linking into the app. I will have to add those to the example app and then check whether an incoming deep-link to the app (not just the Sign in with Apple callback) would pop the browser and show the desired page. This might not be a dealbreaker either way for most consumers of this package, but is something I would want to document clearly (if there is any difference regarding this and the `launchMode`s at all).

**Comment by tp** [comment]
This is now published as v4.3.0: https://pub.dev/packages/sign_in_with_apple
